hat/backends/ptx/src/main/java/hat/backend/PTXCodeBuilder.java

```diff
@@ -0,0 +1,665 @@
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package hat.backend;
+
+import hat.ifacemapper.Schema;
+import hat.optools.*;
+import hat.text.CodeBuilder;
+import hat.util.StreamCounter;
+
+import java.lang.reflect.code.Block;
+import java.lang.reflect.code.Op;
+import java.lang.reflect.code.TypeElement;
+import java.lang.reflect.code.Value;
+import java.lang.reflect.code.op.CoreOp;
+import java.lang.reflect.code.op.ExtendedOp;
+import java.lang.reflect.code.type.JavaType;
+import java.util.*;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+
+public class PTXCodeBuilder extends CodeBuilder<PTXCodeBuilder> {
+
+    Map<Value, PTXRegister> varToRegMap;
+    List<String> params;
+    Map<String, Block.Parameter> paramMap;
+    Map<Field, PTXRegister> fieldToRegMap;
+
+    HashMap<PTXRegister.Type, Integer> ordinalMap;
+
+    PTXRegister returnReg;
+    private int addressSize;
+
+    public enum Field {
+        NTID_X ("ntid.x", false),
+        CTAID_X ("ctaid.x", false),
+        TID_X ("tid.x", false),
+        KC_X ("x", false),
+        KC_ADDR("kc", true),
+        KC_MAXX ("maxX", false);
+
+        private final String name;
+        private final boolean destination;
+
+        Field(String name, boolean destination) {
+            this.name = name;
+            this.destination = destination;
+        }
+        public String toString() {
+            return this.name;
+        }
+        public boolean isDestination() {return this.destination;}
+    }
+
+    public PTXCodeBuilder(int addressSize) {
+        varToRegMap = new HashMap<>();
+        params = new ArrayList<>();
+        fieldToRegMap = new HashMap<>();
+        paramMap = new HashMap<>();
+        ordinalMap = new HashMap<>();
+        this.addressSize = addressSize;
+    }
+
+    public PTXCodeBuilder() {
+        this(32);
+    }
+
+    public void ptxHeader(int major, int minor, String target, int addressSize) {
+        append(".version ").append(String.valueOf(major)).dot().append(String.valueOf(minor)).nl();
+        append(".target ").append(target).nl();
+        append(".address_size ").append(String.valueOf(addressSize));
+        this.addressSize = addressSize;
+    }
+
+    public void functionHeader(String funcName, boolean entry, TypeElement yieldType) {
+        if (entry) {
+            append(".visible .entry ");
+        } else {
+            append(".func ");
+        }
+        if (!yieldType.toString().equals("void")) {
+            returnReg = new PTXRegister(getOrdinal(resultType(yieldType)), resultType(yieldType));
+            returnReg.name("%retReg");
+            oparen().param().space().printParamType(yieldType);
```

Comment:
void preficing with 'print' like this.  everything going through the builder is essentially printing. Maybe just 'paramType(yieldType)'